### DISCUSSION TITLE
Alternative to next-transpile-modules for express app

### DISCUSSION THREAD

**Discussion by NixBiks** [vercel/turborepo#404] [Help] [answered]
Kithen Sink have an example where the nextjs apps uses next-transpile-modules to using internal TS packages without building them. 

What's the alternative for NodeJS server apps? E.g. if the express app in the kitchen sink example was using an internal TS package as well?

Bonus question. Say I have package A and B where B have A as a dependency. I then have my App that depends on B. Would I need my App to explicitly transpile both A and B or is it sufficient to target B?

**Comment by alexn-s** [comment]
I would also love to now.

**Comment by vorillaz** [comment]
@mr-bjerre and @alexn-s Hey there, I am using a similar setup along with turborepo, it is a bit over-engineered but works great for my project.

I am using a centralised `tsconfig.json` for my `next.js` and `express.js` apps, the TypeScript config file exposes my packages and libraries as follows:
```json
{
    "compilerOptions": {
        "baseUrl": ".",
        // ... some other options
        "paths": {
            "@prxm/foo": ["apps/foo/src"],
            "@prxm/bar": ["libs/bar/src"],
            "@prxm/buz": ["libs/buz/src"],
            "@prxm/prisma": ["libs/buz/prisma"]
        }
    }
}
```
Using `paths` i can easily create shortcuts in order to consume each package without linking or hoisting.
For my `next.js` apps `next-transpile-modules` works out of the box.
For my backend applications using express.js I can easily and effectively bundle the entry points using `esbuild` without problems, it's pretty fast as well and you can use additional plugins for node externals, sourcemaps etc.

**Comment by jaredpalmer** [comment, marked as answer]
> Kithen Sink have an example where the nextjs apps uses next-transpile-modules to using internal TS packages without building them.
> 
> What's the alternative for NodeJS server apps? E.g. if the express app in the kitchen sink example was using an internal TS package as well?

An alternative would require you to bundle your Node.js server apps as a build step or on the fly (https://github.com/egoist/esbuild-register like in the Kitchen Sink `api` express app), but also tell that bundler to transpile relevant node_modules correctly. You _must_ ultimately bundle / transpile at some point.  

> Bonus question. Say I have package A and B where B have A as a dependency. I then have my App that depends on B. Would I need my App to explicitly transpile both A and B or is it sufficient to target B?

If you bundle your server code, then yes, this can be possible if you can tell your bundler to bundle all internal dependencies.

**Reply by NixBiks** [reply]
Thanks for the reply. For someone with minimal bundler experience it would still be great with an example. I tend to get confused/overwhelmed by all the tooling around this - I imagine I'm not alone

**Reply by BenJackGill** [reply]
I finally got this working using `tsup` as shown in the [kitchen sink `api` example](https://github.com/vercel/turbo/tree/main/examples/kitchen-sink/apps/api).

You might need a workaround for multiple errors tho. See this for more details and a fix: https://github.com/evanw/esbuild/issues/1921#issuecomment-1575636282 also https://github.com/evanw/esbuild/issues/1921#issuecomment-1898197331
